### ISSUE TITLE
GenericSetup upgradeStep ZCML discriminator should include the profile

### ISSUE THREAD

**Issue by tseaver** [zopefoundation/Products.GenericSetup#7]
In https://bugs.launchpad.net/zope-cmf/+bug/924721, @rpatterson reported:

> It appears the ZCML handler for the upgradeSetp directive doesn't include the profile id in the discriminator. This means that if you register the same handler for two different profiles but with the same source and destination versions, you get a ConfigurationConflictError.

**Comment by rodfersou** [comment]
I get the same problem:
```
Traceback (most recent call last):
  File "/srv/brasil2016.portal/parts/instance/bin/interpreter", line 378, in <module>
    exec(compile(__file__f.read(), __file__, "exec"))
  File "/srv/cache/eggs/Zope2-2.13.24-py2.7.egg/Zope2/Startup/run.py", line 76, in <module>
    run()
  File "/srv/cache/eggs/Zope2-2.13.24-py2.7.egg/Zope2/Startup/run.py", line 22, in run
    starter.prepare()
  File "/srv/cache/eggs/Zope2-2.13.24-py2.7.egg/Zope2/Startup/__init__.py", line 92, in prepare
    self.startZope()
  File "/srv/cache/eggs/Zope2-2.13.24-py2.7.egg/Zope2/Startup/__init__.py", line 268, in startZope
    Zope2.startup()
  File "/srv/cache/eggs/Zope2-2.13.24-py2.7.egg/Zope2/__init__.py", line 47, in startup
    _startup()
  File "/srv/cache/eggs/Zope2-2.13.24-py2.7.egg/Zope2/App/startup.py", line 118, in startup
    load_zcml()
  File "/srv/cache/eggs/Zope2-2.13.24-py2.7.egg/Zope2/App/startup.py", line 52, in load_zcml
    load_site()
  File "/srv/cache/eggs/Zope2-2.13.24-py2.7.egg/Zope2/App/zcml.py", line 46, in load_site
    _context = xmlconfig.file(site_zcml)
  File "/srv/cache/eggs/zope.configuration-3.7.4-py2.7.egg/zope/configuration/xmlconfig.py", line 651, in file
    context.execute_actions()
  File "/srv/cache/eggs/zope.configuration-3.7.4-py2.7.egg/zope/configuration/config.py", line 620, in execute_actions
    for action in resolveConflicts(self.actions):
  File "/srv/cache/eggs/zope.configuration-3.7.4-py2.7.egg/zope/configuration/config.py", line 1527, in resolveConflicts
    raise ConfigurationConflictError(conflicts)
zope.configuration.config.ConfigurationConflictError: Conflicting configuration actions
  For: ('upgradeDepends', '2', '3', None, "[u'plone.app.registry']", 0)
    File "/srv/brasil2016.portal/src/brasil2016/portal/upgrades/v3/configure.zcml", line 22.4-27.10
          <genericsetup:upgradeDepends
              title="Add new tile"
              description="Reload registration to add new tile."
              import_steps="plone.app.registry"
              run_deps="false"
              />
    File "/srv/cache/eggs/collective.texttospeech-1.0rc1-py2.7.egg/collective/texttospeech/upgrades/v3/configure.zcml", line 28.4-33.10
          <genericsetup:upgradeDepends
              title="Add CSS class blacklist field to registry"
              description="Reload registration of configlet registry to add new field."
              import_steps="plone.app.registry"
              run_deps="false"
              />
```

**Comment by mauritsvanrees** [comment]
Confirmed. 
I guess in your case you can fix it by editing `brasil2016/portal/upgrades/v3/configure.zcml` and explicitly adding `import_profile` to the `upgradeDepends` directive.

But say I have this in package A:

```
  <genericsetup:upgradeSteps
      source="1000"
      destination="1001"
      profile="package.a:default">
    <genericsetup:upgradeDepends
        title="Update collection metadata"
        import_profile="package.a:default"
        import_steps="catalog"
        />
  </genericsetup:upgradeSteps>
```

When I have the same in package B and have both `profile` and `import_profile` set to `package.b:default` all is well. But when the `import_profile` is `package.a:default` then I get the same conflict. It probably does not happen very often in practice, but still.

**Comment by rodfersou** [comment]
@mauritsvanrees thank you for your advise.

cc @hvelarde

**Comment by hvelarde** [comment]
@mauritsvanrees we stumbled with this again yesterday; shouldn't this be fixed here in GS? should I open a new issue?

**Comment by tseaver** [comment]
@hvelarde Shouldn't need a new issue, just a patch. :)

**Comment by hvelarde** [comment]
I would love to help, but GS is one of those pieces of software out of my knowledge :-)

**Comment by davisagli** [comment]
@hvelarde if you'd like to take advantage of an opportunity to increase your knowledge, here is the line that is failing to add the profile to the ZCML discriminator: https://github.com/zopefoundation/Products.GenericSetup/blob/master/Products/GenericSetup/zcml.py#L318

**Comment by hvelarde** [comment]
what is a `discriminator`? where and how is used?

**Comment by davisagli** [comment]
The discriminators are used to detect conflicting ZCML registrations. Once a directive is parsed the configuration engine calls the directive handler function I linked above (the specific function for a particular directive). The directive handler registers an "action" which will be called later on (once we're sure there are no conflicts) to actually register components. As part of registering the action it declares a discriminator, which should be whatever uniquely identifies the component that's going to be registered. At this point if there is already an action with the same discriminator, it raises the ConfigurationConflictError

**Comment by davisagli** [comment]
I coincidentally just ran into this bug myself, so I submitted a patch.

**Comment by mauritsvanrees** [comment]
I have released 1.8.8 with David's fix.

**Comment by mauritsvanrees** [comment]
And I released the fix in 1.9.1, for those using Zope2 version 4.
